A7-backward-compatibility-management.md

```diff
@@ -0,0 +1,94 @@
+
+1. Provides a way to identify the gRPC client's versions
+2. Provides a way to manage workarounds on gRPC servers so that they are compatible with previous versions of gRPC clients
+
+## Background
+
+As more versions of gRPC are released to users, different users may use different versions of gRPC at the same time. However these versions of gRPC may not have the same set of features available and may contains bugs to make them behaves incorrectly.
+
+gRPC should be able to do the following:
+* When a gRPC client is of old version and does not support a feature, the server can identify the client's lack of such capability and disable corresponding features appropriately, if needed.
+
+Below use cases have happened multiple times and is expected to continue happening in the future. Known cases are:
+* Compression was not supported in gRPC Objective C with Cronet due to a bug and was fixed in version v1.3.0-dev. The gRPC endpoint on the other side, ESF server in this case, is unable to enable compression only for clients of v1.4.0-dev or higher due to lack of this mechanism being proposed.
+* An old gRPC Objective C client (without Cronet) will crash when getting compressed response.
+* ctiller@ mentioned that there are also some flow control behavior get corrupted due to buggy clients.
+* I remember there are also some compatibililty issue across multiple languages, like the `GO_AWAY` frame issue with golang. We may consult different language teams to add more in this list.
+
+### Related Proposals: 
+N/A
+
+## Proposal
+The major considerations of the design are:
+* Backward compatibility issues should be managed by gRPC server instead of user
+* Servers that do not care about any backward compatibility issue would not need to do anything special
+* Servers that care about a particular backward compatibility issue can invoke an API of gRPC library to configure the server so that the server will apply workaround to that issue based on client's version.
+
+### Client version identification
+gRPC server should identify the version of a client by analysing the `user-agent` header field on the server side. Current format of `user-agent` is well defined by the doc [PROTOCOL-HTTP2](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md). With this definition we can extract gRPC language and version from client's `user-agent` string and identify user's compatibility issues.
+
+### Server backward compatibility management
+gRPC server will maintain a list of server workarounds for backward compatibility issues. These workarounds are disabled by default for performance consideration. If a user cares about a backward compatibility issue, they should be able to enable the workaround. In this case gRPC server will identify the version of each client and determine whether the workaround should be applied.
+
+#### Workaround list
+The workaround list on gRPC server is a map from an integer `workaround id` to a `version comparing function`:
```

Comment:
Avoid talking implementation details here... we almost certainly won't implement it like this (since this API implies that we'd parse user agent strings N times - and that all workarounds would be triggered by user agent - maybe sometimes we'd want to do something else)